File: Extensions/UIKitExtensions/UINavigationController+LXExtension.m
//
//  UINavigationController+LXExtension.m
//
//  Created by 从今以后 on 2017/6/27.
//  Copyright © 2017年 从今以后. All rights reserved.
//

#import "UINavigationController+LXExtension.h"

@implementation UINavigationController (LXExtension)

- (void)_lx_performCompletion:(void (^)(BOOL finished))completion
{
	if (self.transitionCoordinator) {
		[self.transitionCoordinator animateAlongsideTransition:nil completion:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
			!completion ?: completion(!context.isCancelled);
		}];
	} else {
		!completion ?: completion(YES);
	}
}

- (void)lx_pushViewController:(UIViewController *)viewController
					 animated:(BOOL)animated
				   completion:(void (^)(BOOL finished))completion
{
	[self pushViewController:viewController animated:animated];
	[self _lx_performCompletion:completion];
}

- (UIViewController *)lx_popViewControllerAnimated:(BOOL)animated
										completion:(void (^)(BOOL finished))completion
{
	UIViewController *vc = [self popViewControllerAnimated:animated];
	[self _lx_performCompletion:completion];
	return vc;
}

- (NSArray<UIViewController *> *)lx_popToViewController:(UIViewController *)viewController
											   animated:(BOOL)animated
											 completion:(void (^)(BOOL finished))completion
{
	NSArray *vcs = [self popToViewController:viewController animated:animated];
	[self _lx_performCompletion:completion];
	return vcs;
}


- (NSArray<UIViewController *> *)lx_popToRootViewControllerAnimated:(BOOL)animated 
														 completion:(void (^)(BOOL))completion
{
	NSArray *vcs = [self popToRootViewControllerAnimated:animated];
	[self _lx_performCompletion:completion];
	return vcs;
}

@end

File: LXUtilities/UIKitExtensions/Swift/UINavigationController+LXExtension.swift
//
//  UINavigationController+LXExtension.swift
//
//  Created by 从今以后 on 2018/4/16.
//  Copyright © 2018年 从今以后. All rights reserved.
//

import UIKit

extension Swifty where Base: UINavigationController {

	func pushViewController(_ viewController: UIViewController, animated: Bool, completion: @escaping ((Bool)-> Void)) {
		base.pushViewController(viewController, animated: animated)
		performCompletion(completion)
	}

	@discardableResult
	func popViewController(animated: Bool, completion: @escaping ((Bool)-> Void)) -> UIViewController? {
		defer { performCompletion(completion) }
		return base.popViewController(animated: animated)
	}

	@discardableResult
	func popToViewController(
		_ viewController: UIViewController,
		animated: Bool,
		completion: @escaping ((Bool)-> Void))
		-> [UIViewController]?
	{
		defer { performCompletion(completion) }
		return base.popToViewController(viewController, animated: animated)
	}

	@discardableResult
	func popToRootViewController(animated: Bool, completion: @escaping ((Bool) -> Void)) -> [UIViewController]? {
		defer { performCompletion(completion) }
		return base.popToRootViewController(animated: animated)
	}
	
	private func performCompletion(_ completion: @escaping ((Bool) -> Void)) {
		if let transitionCoordinator = base.transitionCoordinator {
			transitionCoordinator.animate(alongsideTransition: nil) { (context) in
				completion(!context.isCancelled)
			}
		} else {
			completion(true)
		}
	}
}
